general/forall.hpp

```diff
@@ -87,58 +87,94 @@ void OmpWrap(const int N, HBODY &&h_body)
 
 
 /// RAJA Cuda backend
-#if defined(MFEM_USE_RAJA) && defined(RAJA_ENABLE_CUDA)
+#if defined(MFEM_USE_RAJA) && defined(RAJA_DEVICE_ACTIVE)
+
+#if defined(RAJA_CUDA_ACTIVE)
+using launch_policy =
+   RAJA::expt::LaunchPolicy<RAJA::expt::null_launch_t, RAJA::expt::cuda_launch_t<false>>;
+using teams_x =
+   RAJA::expt::LoopPolicy<RAJA::loop_exec,RAJA::cuda_block_x_direct>;
+using threads_z =
+   RAJA::expt::LoopPolicy<RAJA::loop_exec,RAJA::cuda_thread_z_direct>;
+#else
+using launch_policy =
+   RAJA::expt::LaunchPolicy<RAJA::expt::null_launch_t, RAJA::expt::hip_launch_t<false>>;
+using teams_x =
+   RAJA::expt::LoopPolicy<RAJA::loop_exec,RAJA::hip_block_x_direct>;
+using threads_z =
+   RAJA::expt::LoopPolicy<RAJA::loop_exec,RAJA::hip_thread_z_direct>;
+#endif
 
-using RAJA::statement::Segs;
 
 template <const int BLOCKS = MFEM_CUDA_BLOCKS, typename DBODY>
-void RajaCudaWrap1D(const int N, DBODY &&d_body)
+void RajaDeviceWrap1D(const int N, DBODY &&d_body)
 {
    //true denotes asynchronous kernel
-   RAJA::forall<RAJA::cuda_exec<BLOCKS,true>>(RAJA::RangeSegment(0,N),d_body);
+#if defined(RAJA_CUDA_ACTIVE)
+   using ForPolicy = RAJA::cuda_exec<MFEM_CUDA_BLOCKS,true>;
+#else
+   using ForPolicy = RAJA::hip_exec<MFEM_CUDA_BLOCKS,true>;
+#endif
+   RAJA::forall<ForPolicy>(RAJA::RangeSegment(0,N),d_body);
 }
 
 template <typename DBODY>
-void RajaCudaWrap2D(const int N, DBODY &&d_body,
-                    const int X, const int Y, const int BZ)
+void RajaDeviceWrap2D(const int N, DBODY &&d_body,
+                      const int X, const int Y, const int BZ)
 {
    MFEM_VERIFY(N>0, "");
    MFEM_VERIFY(BZ>0, "");
    const int G = (N+BZ-1)/BZ;
-   RAJA::kernel<RAJA::KernelPolicy<
-   RAJA::statement::CudaKernelAsync<
-   RAJA::statement::For<0, RAJA::cuda_block_x_direct,
-        RAJA::statement::For<1, RAJA::cuda_thread_x_direct,
-        RAJA::statement::For<2, RAJA::cuda_thread_y_direct,
-        RAJA::statement::For<3, RAJA::cuda_thread_z_direct,
-        RAJA::statement::Lambda<0, Segs<0>>>>>>>>>
-        (RAJA::make_tuple(RAJA::RangeSegment(0,G), RAJA::RangeSegment(0,X),
-                          RAJA::RangeSegment(0,Y), RAJA::RangeSegment(0,BZ)),
-         [=] RAJA_DEVICE (const int n)
+
```

Comment:
maybe add `using Raja::expt;` here.

Reply:
> I'm not sure I understand the benefit of `Teams` over the current approach, is it simpler to read? Otherwise, this looks good, I need to run on AMD GPUs to test it though.

The main improvement is that we have been able to get improved performance over the RAJA::kernel API for certain MFEM kernels. Teams in more light weight and gives more direct access to setting up the compute grid on gpus.  I'll clean this up today.

Reply:
Done! Only the items related to RAJA Teams is under the ::expt name space.